compiler/packages/babel-plugin-react-compiler/src/HIR/CollectHoistablePropertyLoads.ts

```diff
@@ -0,0 +1,520 @@
+      /**
+       * Since this runs *after* buildReactiveScopeTerminals, identifier mutable ranges
+       * are not valid with respect to current instruction id numbering.
+       * We use attached reactive scope ranges as a proxy for mutable range, but this
+       * is an overestimate as (1) scope ranges merge and align to form valid program
+       * blocks and (2) passes like MemoizeFbtAndMacroOperands may assign scopes to
+       * non-mutable identifiers.
+       *
+       * See comment at top of function for why we track known immutable identifiers.
+       */
+      const isMutableAtInstr =
+        object.mutableRange.end > object.mutableRange.start + 1 &&
+        object.scope != null &&
+        inRange(instr, object.scope.range);
+      if (
+        !isMutableAtInstr ||
+        knownImmutableIdentifiers.has(propertyNode.fullPath.identifier)
+      ) {
+        let curr = propertyNode.parent;
+        while (curr != null) {
+          assumedNonNullObjects.add(curr);
+          curr = curr.parent;
+        }
+      }
+    }
+
+    nodes.set(block.id, {
+      block,
+      assumedNonNullObjects,
+    });
+  }
+  return nodes;
+}
+
+function deriveNonNull(
+  fn: HIRFunction,
+  nodes: ReadonlyMap<BlockId, BlockInfo>,
+): void {
+  const succ = new Map<BlockId, Set<BlockId>>();
+  const terminalPreds = new Set<BlockId>();
+
+  for (const [blockId, block] of fn.body.blocks) {
+    for (const pred of block.preds) {
+      getOrInsertDefault(succ, pred, new Set()).add(blockId);
+    }
+    if (block.terminal.kind === 'throw' || block.terminal.kind === 'return') {
+      terminalPreds.add(blockId);
+    }
+  }
+
+  function recursivelyDeriveNonNull(
+    nodeId: BlockId,
+    direction: 'forward' | 'backward',
+    traversalState: Map<BlockId, 'active' | 'done'>,
+    nonNullObjectsByBlock: Map<BlockId, Set<PropertyLoadNode>>,
+  ): boolean {
+    /**
+     * Avoid re-visiting computed or currently active nodes, which can
+     * occur when the control flow graph has backedges.
+     */
+    if (traversalState.has(nodeId)) {
+      return false;
+    }
+    traversalState.set(nodeId, 'active');
+
+    const node = nodes.get(nodeId);
+    if (node == null) {
+      CompilerError.invariant(false, {
+        reason: `Bad node ${nodeId}, kind: ${direction}`,
+        loc: GeneratedSource,
+      });
+    }
+    const neighbors = Array.from(
+      direction === 'backward' ? (succ.get(nodeId) ?? []) : node.block.preds,
+    );
+
+    let changed = false;
+    for (const pred of neighbors) {
+      if (!traversalState.has(pred)) {
+        const neighborChanged = recursivelyDeriveNonNull(
+          pred,
+          direction,
+          traversalState,
+          nonNullObjectsByBlock,
+        );
+        changed ||= neighborChanged;
+      }
+    }
+    /**
+     * Active neighbors can be filtered out as we're solving for the following
+     * relation.
+     * X = Intersect(X_neighbors, X)
```

Comment:
This doesn't look like it describes the algorithm. It looks like what we're actually getting is Union(X, Intersect(...X_neighbors)), no? (cf: Set_intersect on line 422, and Set_union on line 433)

Reply:
Yep that's totally right, I separated this comment into two -- one describing the general algorithm (the Union(X, Intersect(...)) and another one describing the active neighbors edge case specifically.

Also updated PR summary to be correct. Thanks for calling this out!